### PR TITLE
[DCJ-653] Add upgrade flight to set Autoclass ARCHIVE on a bucket

### PR DESCRIPTION
**Pull request by fboulnois**
## Addresses

https://broadworkbench.atlassian.net/browse/DCJ-653

## Summary of changes

Runs an upgrade flight which:

- Checks and records the current actual bucket Autoclass state
- Sets the bucket Autoclass state to ARCHIVE, or restores the previous one
- Sets the bucket Autoclass metadata state to true, or restores the previous one

## Testing Strategy

Unit test coverage of all the main components, and manual testing using a local TDR instance against a dev bucket. The upgrade service can be called like this:

```
{
  "upgradeName": "20240911_bucket_upgrade",
  "upgradeType": "custom",
  "customName": "SET_BUCKET_AUTOCLASS_ARCHIVE",
  "customArgs": [
    "mybucketname"
  ]
}
```

Before:

```
$ gcloud storage buckets describe gs://mybucketname --format=json
{
...
  "default_storage_class": "REGIONAL",
...
}
```

After:

```
$ gcloud storage buckets describe gs://mybucketname --format=json
{
...
  "autoclass": {
    "enabled": true,
    "terminalStorageClass": "ARCHIVE",
    "terminalStorageClassUpdateTime": "2024-09-12T00:32:03.001000+00:00",
    "toggleTime": "2024-09-12T00:32:03.001000+00:00"
  },
  "autoclass_enabled_time": "2024-09-12T00:32:03+0000",
  "default_storage_class": "STANDARD",
...
}
```

### CODE REVIEW

File: src/main/java/bio/terra/common/ValidationUtils.java [hunk 1]
```diff
@@ -72,6 +76,17 @@ public static Optional<UUID> convertToUuid(String uuid) {
     }
   }
 
+  public static boolean isValidBucketName(String bucketName) {
+    return VALID_BUCKET_NAME_REGEX.matcher(bucketName).matches();
+  }
+
+  public static Optional<String> convertToBucketName(String bucketName) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/common/ValidationUtils.java:83]: Since this method doesn't modify its argument, I don't see what the purpose of it is. Instead I would only have `isValidBucketName()`, and call it like this:
```
          String bucketName = request.getCustomArgs().get(0);
          Preconditions.checkArgument(
              ValidationUtils.isValidBucketName(bucketName), "Custom argument's single value is not a valid bucket name");

          return Map.of(JobMapKeys.BUCKET_NAME.getKeyName(), bucketName);
```

The extra work to make and unwrap an `Optional` is more verbose and doesn't add any clarity to the code IMO.

File: src/test/java/bio/terra/common/ValidationUtilsTest.java [hunk 2]
```diff
@@ -58,6 +58,40 @@ void testUuidFormats() {
     assertThat(ValidationUtils.convertToUuid("not a uuid")).isEmpty();
   }
 
+  @Test
+  void testIsValidBucketName() {
+    assertThat(ValidationUtils.isValidBucketName("bucket")).isTrue();
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/common/ValidationUtilsTest.java:63]: It is odd that this code is using assertj `assertThat()` instead of hamcrest. In general we don't use `assertj` APIs so this code should be migrated at some point.

File: src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    FlightMap workingMap = context.getWorkingMap();
+    GoogleBucketResource bucketResource = googleBucketService.getBucketMetadata(bucketName);
+    Bucket bucket = googleBucketService.getCloudBucket(bucketName);
+    bucketResource = bucketResource.storageClass(bucket.getStorageClass());
+    Autoclass autoclass = bucket.getAutoclass();
+    if (autoclass != null) {
+      if (autoclass.getEnabled() && autoclass.getTerminalStorageClass() == StorageClass.ARCHIVE) {
+        return stepResultFailure("Bucket autoclass already set to ARCHIVE: " + bucketName);
+      }
+      if (autoclass.getEnabled() ^ bucketResource.getAutoclassEnabled()) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java:42]: Instead of `^`, would `!=` work? That would be clearer if so
```suggestion
      if (autoclass.getEnabled() != bucketResource.getAutoclassEnabled()) {
```

File: src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException {
+    FlightMap workingMap = context.getWorkingMap();
+    GoogleBucketResource bucketResource = googleBucketService.getBucketMetadata(bucketName);
+    Bucket bucket = googleBucketService.getCloudBucket(bucketName);
+    bucketResource = bucketResource.storageClass(bucket.getStorageClass());
+    Autoclass autoclass = bucket.getAutoclass();
+    if (autoclass != null) {
+      if (autoclass.getEnabled() && autoclass.getTerminalStorageClass() == StorageClass.ARCHIVE) {
+        return stepResultFailure("Bucket autoclass already set to ARCHIVE: " + bucketName);
```
Comment (rushtong) [inline src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java:40]: Does this have to be a failure or is there a kind of step result that says, ignored for reason X?

Comment (fboulnois) [reply src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java:40]: I think it has to be a failure, otherwise the other steps will try to run and additional checks will be needed in each subsequent step.

**Review by rushtong — APPROVED** [review]
This looks reasonable to me 👍🏽

File: src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketResource.java [hunk 5]
```diff
@@ -78,6 +81,53 @@ public GoogleBucketResource autoclassEnabled(boolean autoclassEnabled) {
...
+
+  // Required to avoid serialization/deserialization issues with Stairway
+  public GoogleBucketResource storageClassAsString(String storageClassAsString) {
+    this.storageClassAsString = storageClassAsString;
+    return this;
+  }
+
+  @JsonIgnore
+  public StorageClass getStorageClass() {
+    return storageClassAsString == null ? null : StorageClass.valueOf(storageClassAsString);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketResource.java:97]: What effect does using `null` have here? If `null` is passed to `GoogleBucketService.setBucketAutoclass()`, will that work?

If the google APIs don't support null, then I think this object shouldn't support it either.

File: src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      return stepResultFailure("Bucket not found: " + bucketName);
+    }
+    Autoclass autoclass = bucket.getAutoclass();
+    if (autoclass.getEnabled() && autoclass.getTerminalStorageClass() == StorageClass.ARCHIVE) {
+      return stepResultFailure("Bucket autoclass already set to ARCHIVE: " + bucketName);
+    }
+    if (autoclass.getEnabled() ^ bucketResource.getAutoclassEnabled()) {
+      return stepResultFailure("Bucket autoclass mismatch in metadata: " + bucketName);
+    }
+    if (autoclass.getEnabled()) {
```
Comment (rjohanek) [inline src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java]: autoclass getEnabled() could be a variable maybe. Also, can you explain this step again, we want to record the current state in case of a rollback right? why fail if the bucket is already set to archive? why have a separate case for if autoclass is enabled compared to if it isn't - what is being recorded differently?

Comment (fboulnois) [reply src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java]: We don't _have_ to fail if the bucket is Autoclass is set to ARCHIVE, but if we don't we end up applying the same policies again to the bucket, so I prefer to fail fast instead.

As for your other question, the problem is that Google changed how the Autoclass behavior works in October of 2023. Before it was the case that Autoclass terminated at ARCHIVE automatically, but now Autoclass terminates at NEARLINE instead. So it's not enough to simply rollback the Autoclass state, but the terminal storage class must also be recorded to be able to rollback as well.

In other words, we need to store the current class of the bucket, whether Autoclass is turned on, and what the Autoclass terminal class is. All three of these must be restored to their previous values in case of rollback.

**Review by rjohanek — APPROVED** [review]
this looks good to me, thanks!

**Review by snf2ye — APPROVED** [review]
I followed up with one question offline, but this looks reasonable to me.

File: src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    return new StepResult(
+        StepStatus.STEP_RESULT_FAILURE_FATAL, new GoogleResourceException(message));
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException {
+    FlightMap workingMap = context.getWorkingMap();
+    GoogleBucketResource bucketResource = googleBucketService.getBucketMetadata(bucketName);
+    Bucket bucket = googleBucketService.getCloudBucket(bucketName);
+    bucketResource = bucketResource.storageClass(bucket.getStorageClass());
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java:36]: This is a bit confusing as `storageClass()` is a fluent-style setter but this isn't a fluent context. Instead I would add a non-fluent setter to use here.
```suggestion
    bucketResource.setStorageClass(bucket.getStorageClass());
```

Comment (fboulnois) [reply src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java:36]: You're right, this is an oversight on my part.

File: src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java (hunk 3 above)
Comment (fboulnois) [reply src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java:42]: I think so, the XOR conditions need to be true:

| L | R | result |
| --- | --- | --- |
| true | true | false |
| false | true | true |
| true | false | true |
| false | false | false |

File: src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketResource.java (hunk 5 above)
Comment (fboulnois) [reply src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketResource.java:97]: It is possible that the terminal StorageClass is null. It will work if `Autoclass.SetEnabled` is set to false.
